File: Algorithms/Sorting Algorithms/Quick Sort/Quicksort.py
# quick sort
def partition(myList, start, end):
    pivot = myList[start]
    left = start+1
    # Start outside the area to be partitioned
    right = end
    done = False
    while not done:
        while left <= right and myList[left] <= pivot:
            left = left + 1
        while myList[right] >= pivot and right >=left:
            right = right - 1
        if right < left:
            done = True
        else:
            # swap places
            temp = myList[left]
            myList[left] = myList[right]
            myList[right] = temp
    # start with myList[right]
    temp = myList[start]
    myList[start] = myList[right]
    myList[right] = temp
    return right


def quick_sort(myList, start, end):
    if start < end:
        # partitioning the list
        split = partition(myList, start, end)
        # split halves & sort both halves
        quick_sort(myList, start, split - 1)
        quick_sort(myList, split + 1, end)
    return myList



my_List = [7,2,5,1,29,6,4,19,11]
sortedList = quick_sort(my_List, 0, len(my_List) - 1)
print(sortedList)

File: Algorithms/Sorting Algorithms/Quick Sort/Java/QuickSort.java
import java.util.ArrayList;
import java.util.Collections;

public class QuickSort<T extends Comparable<T>> 
{
  private int partition(ArrayList<T> arrayList, int start, int end) {
    T pivot = arrayList.get(start);
    boolean done = false;
    int original = start;
    start += 1;
    while(!done) {
      // Iterate until "start" points to a value that is bigger than the pivot
      while (arrayList.get(start).compareTo( pivot) < 0 && start < end) {
        start += 1;
      }
      // Iterate until "end" points to a value that is smaller than the pivot
      while (arrayList.get(end).compareTo(pivot) >= 0 && start < end) {
        end -= 1;
      }
      // Swap them
      if (start < end) {
        Collections.swap(arrayList, start, end);
        start+=1;
      } else {
        done = true;
      }
  
    }

    // The Loop will always end with start=end. We must check if that position is smaller or bigger than the pivot
    if (arrayList.get(start).compareTo( pivot) > 0) {
	start -= 1;
    }
    // Put pivot in the middle
    Collections.swap(arrayList, original, start);
    return start;
  }
  
  private int partition(T[] arrayList, int start, int end) {
    T pivot = arrayList[start];
    boolean done = false;
    int original = start;
    start += 1;
    while(!done) {
      // Iterate until "start" points to a value that is bigger than the pivot
      while (arrayList[start].compareTo( pivot) < 0 && start < end) {
        start += 1;
      }
      // Iterate until "end" points to a value that is smaller than the pivot
      while (arrayList[end].compareTo(pivot) >= 0 && start < end) {
        end -= 1;
      }
      // Swap them
      if (start < end) {
        T temp = arrayList[start];
        arrayList[start] = arrayList[end];
        arrayList[end] = temp;
        start+=1;
      } else {
        done = true;
      }

    }

    // The Loop will always end with start=end. We must check if that position is smaller or bigger than the pivot
    if (arrayList[start].compareTo( pivot) > 0) {
      start -= 1;
    }
    // Put pivot in the middle
    arrayList[original] = arrayList[start];
    arrayList[start] = pivot;
    return start;

  }

  public void sort(ArrayList<T> arrayList) {
	sort(arrayList, 0, arrayList.size()-1);
  }
                      
  public void sort(ArrayList<T> arrayList, int start, int end) {
    if (start < end) {
      int split = partition(arrayList, start, end);
      sort(arrayList, start, split-1);
      sort(arrayList, split+1, end);
    }
  }

  public void sort(T[] array) {
	sort(array, 0, array.length-1);
  }
  
  public void sort(T[] arrayList, int start, int end) {
    if (start < end) {
      int split = partition(arrayList, start, end);
      sort(arrayList, start, split-1);
      sort(arrayList, split+1, end);
    }
  }
}
